crates/web-plugins/didcomm-messaging/protocols/forward/src/handler.rs

```diff
@@ -36,15 +43,38 @@ pub(crate) async fn mediator_forward_process(
             AttachmentData::Base64 { value: data } => json!(data.base64),
             AttachmentData::Links { value: data } => json!(data.links),
         };
-        message_repository
-            .store(RoutedMessage {
-                id: None,
-                message: attached,
-                recipient_did: next.as_ref().unwrap().to_owned(),
-            })
-            .await
-            .map_err(|_| ForwardError::InternalServerError)?;
+
+        let result = retry_async(
+            || {
+                let attached = attached.clone();
+                let recipient_did = next.as_ref().unwrap().to_owned();
+
+                async move {
+                    message_repository
+                        .store(RoutedMessage {
+                            id: None,
+                            message: attached,
+                            recipient_did,
+                        })
+                        .await
+                }
+            },
+            RetryOptions::new()
+                .retries(5)
+                .exponential_backoff(Duration::from_millis(100))
+                .max_delay(Duration::from_secs(1)),
+        )
+        .await;
```

Comment:
Don't you see you're repeating this logic over and over? Since you're making the repository layer resilient, I believe you should create an overloaded base repository with the retry logic, making it being reused transparently. Note also that some RepositoryError's should not lead to a retry like you're assuming here. WDYT?